### ISSUE TITLE
PUB - SUB sockets randomly stop working?

### ISSUE THREAD

**Issue by vreascul** [zeromq/jeromq#364]
Hello,

We are using ZMQ for our server (reactphp/zmq) and our client (zeromq/jeromq). We run multiple instances of the server and each server has several clients connecting to it. We use several other sockets but we are facing an issue with the TCP communication between the PUB socket on the server and the SUB ones on the clients.
The issue is very odd as it is not easy to replicate but from time to time the client stops receiving messages from the server. This happens on all of our server instances and their clients once every 2-3 months but we have a new server instance that replicates this issue in around 20 - 30 minutes. Our servers are "identical" and this points us to a network difference.

We made a simple app that just opens several threads with SUB sockets and listen on the same topic. They all fail eventually but we observed that not all of them fail at the same time. Also, restarting a client solves the issue so we think that the issue is most likely with the client.
The server logs indicate that the send is called on the PUB socket and the listening thread is still running on the client calling recvStr on the SUB socket once every several seconds. This points us to a ZMQ issue but we don't know it for sure.

In any case, right now we are looking into the jeromq implementation but we are wondering if anyone else experienced something like this and maybe has some pointers. 

Thank you.

**Comment by galves** [comment]
Are you using TCP and on what OS?  Note that socket.setTCPKeepAliveInterval & setTCPKeepAlive don't work on JeroMQ so do not rely on them to reconnect the socket after network loss.

You might need to enable tcp connection polling on your client/server and see if that helps. e.g.

 # echo 600 > /proc/sys/net/ipv4/tcp_keepalive_time

  # echo 60 > /proc/sys/net/ipv4/tcp_keepalive_intvl

  # echo 20 > /proc/sys/net/ipv4/tcp_keepalive_probes

**Comment by andreaslarssen** [comment]
I work with @vreascul on this project. For now we run it on Windows (7, 8 and 10), as well as a Raspbian. Should run on Ubuntu and OSX in the future as well. 

Thanks for the tip, @galves, we'll look into it and come back with the findings.

**Comment by b6arat6** [comment]
@vreascul, @andreaslarssen We also face the same issue in our production, when we migrated to jeromq.

Did you guys get any idea/finding to solve the issue?

**Comment by andreaslarssen** [comment]
Hi @b6arat6. We're experimenting with tcp config settings, but it hasn't solved anything yet. If that indeed solves the issue, it's still kinda troubling for us as we have to config tcp connections in install scripts (we're building an install-able client). 

I tried at an earlier stage to heartbeat from server to client and reconnect on missed heartbeats, but that too failed (that off course might have been my implementation that failed as it seems like a straight forward way to achieve a steady connection). After that, @vreascul  took over, and he has tried to listen for disconnect events, but with no luck. 

Next step might be to re-try the server->client heartbeating or use zmq java bindings instead of jeromq, even though we have to download and install zmq with our client installer. 

Did you guys just get aware of the issue, or have you puzzled with this for a while? I'm really interested in what you've tried to solve it so we don't have to go down that same road. And also how this issue manifests itself. For us, one server -> clients connection works (almost) like a charm, while the other (identical setup) fails every X minutes / hours. When I say identical, I'm not counting the router / firewall between the server -> clients, as well as the clients network. While I don't think there are any "jeromq hating" configurations in the firewall or network, an unstable network for example might trigger the issue at one place, not the other. 

Atm. @vreascul is working on this and digging into the jeromq source, so he might have some details if you have any specific questions.

**Comment by andreaslarssen** [comment]
Also, there must be someone out there running the PUB/SUB pattern with success? Someone who can give an example of an implementation that works? How are people handling sub socket disconnects if jeromq is not?

**Comment by colini** [comment]
We've been running jeromq 0.3.4 and 0.3.5 PUB/SUB for several years on Ubuntu and it has worked fine.  SUB sockets connect to the PUB.  We do send a heartbeat every second if no data has been published.  No TCP kernel tweaks.

Maybe try a packet sniffer to see what's happening on the wire?

**Comment by andreaslarssen** [comment]
No tcp disconnects on the sub side? How do you handle hearbeats on the sub side?

**Comment by colini** [comment]
No disconnects that I'm aware of.  The SUB clients don't do anything with the heartbeat message.  It may not even be necessary.

**Comment by andreaslarssen** [comment]
Does that mean the heartbeat is necessary to keep the pub socket open? If you're not using it do discover disconnects?

**Comment by andreaslarssen** [comment]
@vreascul tried setting both the keeptime and the interval for tcp sockets (in windows), but it has no effect on the issue. Next step is trying the pub heartbeat again.

**Comment by colini** [comment]
TBH, it was the first zeromq application I had written and I don't remember why I put the hb there.  Looking at the code, the SUB client does not use it to detect disconnects or reconnect, etc.  It just receives the hb and ignores it.  

My understanding is that connect/reconnect is one of zeromq's main features -- it's supposed to take care of that for you.

If you're still stuck I recommend sniffing the traffic on the network.  This is a good article that explains how to do that with wireshark: http://www.codeproject.com/Articles/863889/ZeroMQ-Diving-Into-The-Wire

**Comment by andreaslarssen** [comment]
We've used wireshark at an earlier point to confirm that the the traffic is actually sent, and it is. 

I was also under the impression that connect/reconnect was one of the zmq features, but it seems like that's not the case: http://stackoverflow.com/questions/26112992/zeromq-pub-sub-on-unreliable-connection. They also suggest the heartbeating, so that's what we'll try (again) next,

**Comment by b6arat6** [comment]
@andreaslarssen @vreascul Our system has n server (n-PUB), n clients (n-SUB). n-SUB have subscribed to n-PUB. The system had been working without any issues related to message loss until we migrated to jeromq-0.3.2 (from maven repo) from zmq.jar (directly, not from maven repo) along with native libs, just a week ago.

Of n-SUB, 1-SUB stops receiving messages from 1-PUB after sometime (did not note the interval), whereas (n-1)-SUB continues to receive messages from the 1-PUB.

We already have heart beat mechanism in place - PUB sends a heartbeat at every 10 mins to n-SUB. If SUB has missed a heartbeat in last 10 mins from any PUB, [ it closes the socket with org.zeromq.ZMQ.Socket.close(), re-connects (i.e. connects to n-PUB all over again with org.zeromq.ZMQ.Socket.connect(String addr)) ] - Action1.

When the issue happens, neither message nor heartbeat is received at the SUB. Even though Action1 is taken by the SUB, the SUB does not receive anything from the PUB.

The issue was solved, after 1. restarting the SUB, 2. restarting the PUB.

We are yet to enable TCP connection polling.

Our VMs has,
tcp_keepalive_time = 7200
tcp_keepalive_intvl = 75
Tcp_keepalive_probes = 9

Will 0mq at SUB side identify the connection broken state based on tcp config & reconnect itself? Is it what called as tcp config in Linux helping/aiding 0mq to recover in such instances?

Wireshark - We are yet to try.

I will keep posting our progress.

**Comment by andreaslarssen** [comment]
@b6arat6:

We're using 0.3.5 via Maven. 
About the connection broken state, I would have to ask @vreascul.

**Comment by vreascul** [comment]
Hello,

We are monitoring the SUB socket (on all events) and we can't see any event even if we disconnect the network. I think that this might by design as PUB and SUB sockets seem to be made for really fast one way communication and probably they had to drop any extra chattering for the performance. We did not use Wireshark so far and i have a limited understanding of the underlying socket (java.nio.channels.SocketChannel) that jeromq is using for SUB so i can not tell for sure that this is the case but it is my best guess.

For us, manually restarting the client (the SUB socket) solves the issue so we just need to find out when the issue happens and restart the SUB socket. For this we are looking once again into sending pings and , hopefully, finding a way to restart the SUB socket successfully (we tried this approach in the past and it did not work but i'm not sure what happened as i was not involved in this project at that time).

As for using OS settings for the sockets keepalive, we tried this in windows with no success. Not sure how these settings are supposed to interact with the java.nio.channels.SocketChannel but the fact that it did not fix the issue points us in these directions:
1. the issue is not with the socket (since we can not see the socket failing we can not really rule this out but it is very unlikely)
2. The SocketChannel is not really using the "thing" that the windows settings are for.
3. I messed up setting those register values up (https://technet.microsoft.com/en-us/library/cc957549.aspx and  https://technet.microsoft.com/en-us/library/cc957548.aspx)

Not sure if this answers you question about the broken state but if you need something more specific please let me know.

Regards!

**Comment by galves** [comment]
I've been using Jeromq 3.3 in production systems for 2-3 years now.  We handle many thousands of msg/sec without issue and there has been no imperative to switch to jzmq.  Before enabling TCP probes on the client however, we had the same issue that you describe...

Our use case is strictly server-sever messaging though and most likely we would consider a different solution for connecting end user clients - as you seem to be doing.

```
# the interval between the last data packet sent (simple ACKs are not considered data) and the first keepalive probe; after the connection is marked to need keepalive, this counter is not used any further
tcp_keepalive_time = 7200

# the interval between subsequential keepalive probes, regardless of what the connection has exchanged in the meantime
tcp_keepalive_intvl = 75

# the number of unacknowledged probes to send before considering the connection dead and notifying the application layer
Tcp_keepalive_probes = 9
```

So in your scenario, you would need to wait 7200 + (75 * 9) seconds before the **server** marks the connection as dead.  However it is more likely the **client** having issues.  The Client does auto reconnect as you expect, but only once the underlying socket has been closed.  If the connectivity is lost but the socket remains open then the socket will never reconnect.  In Linux you can quite easily simulate this with iptables - set it to silently discard traffic and check netstat to see if the socket remains open.  You can probably do something similar in Windows - or just unplug your network cable.

Try setting tcp_keepalive_time = 60, Tcp_keepalive_probes = 1 on the client and see if that helps.

**Comment by b6arat6** [comment]
@vreascul @galves Yes, manually restarting SUB, solves the issue but it keeps coming. We could not try wireshark, as its a production machine & enabling connection pooling, as it requires a machine rebooting. 

We have planned to fall back to zmq.jar (3.x) along with native libraries for one of our regions & see whether the issue stops occurring.

We use, void org.zeromq.ZMQ.Socket.close() which internally calls protected native void destroy();.

**Comment by daveyarwood** [comment]
If restarting the SUB socket works, then could you do that regularly
instead of waiting for the problem to happen? Say, restart every 5 minutes
or something like that?

**Comment by andreaslarssen** [comment]
Still researching this, but heartbeating from pub definitely had an impact. 

We also implemented reconnecting of sub sockets on one client to compare to clients without this feat, but still not sure if it has an impact or not. Both seems to work most of the time. Still trying to figure out what the occasional errors are caused by.  Might not be related to zmq / jeromq at all, will get back.

**Comment by luxor99** [comment]
I have a very simple Java Pub/Nodejs Sub setup and I thought I was going crazy with the connection going down.  Now I see other people also have the issue.  The Publisher would eventually get into a state where even restarting the Subscriber would not allow it to reconnect.  I switched to jzmq and now have no problems.  It's too bad, a pure java implementation of zeromq seemed promising, but I'm not looking back.

**Comment by vtellier** [comment]
I'm using a set of publisher and subscriber using libzmq C library. They all connect to a NodeJs proxy (that binds zeromq 4.2.2)

I had this problem where the sucribers silently stop to get messages from the publisher, especially when the publisher didn't published anything for a while (~20 minutes).

The ZMQ heartbeat fixed it quite well so far.

**Comment by andreaslarssen** [comment]
We too have to rewrite this i another language implementation or lib. Can't seem to get it working-
Are the lib owners / contributers aware of the situation (again, if it's not us doing something completely bonkers); @trevorbernard, @miniway etc?

**Comment by andreaslarssen** [comment]
@vtellier: Did you post something you deleted?

**Comment by vtellier** [comment]
Yes, I realized posted on the wrong project just after having posted... we are here on jeromq not libzmq.

My post mentionned that _I face a similar behavior and the ZMQ heartbeat seems to be an effective workaround for me as well._
I do use the C library, not jeromq. It could mean the issue is related, the thing is that I know nothing about jeromq, therefor I deleted my post. I'll see if there is another issue in libzmq project.

**Comment by arvindNeergunda** [comment]
I am also seeing this issue. From the looks of it, the PUB seems to be publishing fine, but the 3 subscribers just seem to not get it randomly and once they stop getting, it continues to stay in that state. 
jeromq version : version:'0.4.0'
Restart of the subs gets them working. The publisher process was not restarted. Are there any specific logs that could help out here. ? I am ready to do some digging around, some pointers on where to look into would be helpful. 
NOTE:  I don't have any great amount of traffic on this pub/sub socks. It randomly just stops working. Sometimes its after 40 days, sometimes its after 4 days. Makes it very indeterministic.
Was looking into jzmq, and it also seems to have some issue 
@b6arat6  seems to have reported an issue in pub/sub on jzmq ( https://github.com/zeromq/jzmq/issues/463 ). Is it for trying the same usecase on jzmq ??

**Comment by daveyarwood** [comment]
As several people have mentioned, heartbeating and/or restarting nodes is an effective way of keeping the connection open. My impression is that this is an inherent issue with ZeroMQ in general, possibly even part of the design (i.e. maybe ZeroMQ makes no attempt to keep the connection open, and it's on us to design our architecture such that the connections stay in use / open).
